### PR TITLE
[sqlserver] The first draft of sqlserver-cdc

### PR DESCRIPTION
**Pull request by GOODBOY008**
The first draft that offers sqlserver-cdc to capture sqlserver table changes.

### CODE REVIEW

**Comment by GOODBOY008** [comment]
When all works done, I will request CR.

**Comment by GOODBOY008** [comment]
@leonardBang Please code review.

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/StartupOptions.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+     * from now onwards should be propagated to topics.
+     */
+    public static StartupOptions schemaOnly() {
+        return new StartupOptions(StartupMode.SCHEMA_ONLY);
+    }
+
+    private StartupOptions(StartupMode startupMode) {
+        this.startupMode = startupMode;
+
+        switch (startupMode) {
```
Comment (hektor8603) [inline flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/StartupOptions.java:54]: add  specific_offsets from lsn

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/SqlServerSource.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            props.setProperty("database.password", checkNotNull(password));
+            props.setProperty("database.port", String.valueOf(port));
+            props.setProperty("database.history.skip.unparseable.ddl", String.valueOf(true));
+            props.setProperty("database.dbname", checkNotNull(database));
+
+            if (tableList != null) {
+                props.setProperty("table.include.list", String.join(",", tableList));
+            }
+
+            switch (startupOptions.startupMode) {
```
Comment (equalsX) [inline flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/SqlServerSource.java:133]: add specific offsets from lsn ?

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableSource.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                        .setPhysicalRowType(physicalDataType)
+                        .setMetadataConverters(metadataConverters)
+                        .setResultTypeInfo(typeInfo)
+                        .build();
+        DebeziumSourceFunction<RowData> sourceFunction =
+                SqlServerSource.<RowData>builder()
+                        .hostname(hostname)
+                        .port(port)
+                        .database(database)
+                        .tableList("dbo." + tableName)
```
Comment (ashulin) [inline flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableSource.java:124]: If the user is not the default schema name `dbo` ?

**Review by ashulin — COMMENTED** [review]
It seems that the code for the startup mode on the table api is still missing.
Looking forward to the completion of the PR

**Comment by GOODBOY008** [comment]
> It seems that the code for the startup mode on the table api is still missing. Looking forward to the completion of the PR



> </dependency>

@ashulin Thank you for your code review, I will fix soon.

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/SqlServerSource.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                    break;
+                default:
+                    throw new UnsupportedOperationException();
+            }
+
+            if (dbzProperties != null) {
+                dbzProperties.forEach(props::put);
+            }
+
+            return new DebeziumSourceFunction<>(
```
Comment (equalsX) [inline flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/SqlServerSource.java:151]: Add a validator to verify whether the database version matches and whether CDC is enabled?

Comment (GOODBOY008) [reply flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/SqlServerSource.java:151]: > Add a validator to verify whether the database version matches and whether CDC is enabled?

Ok,It should  add cdc enabled validation. Thanks for your advice.

**Comment by GOODBOY008** [comment]
I found sqlserver-cdc connector should add  e2e test.I will add this test.

**Comment by leonardBang** [comment]
> I found sqlserver-cdc connector should add e2e test.I will add this test.

@GOODBOY008 Nice catch, you can create an issue firstly and then open the PR.

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableFactory.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription("Database name of the SqlServer server to monitor.");
+
+    private static final ConfigOption<String> TABLE_NAME =
+            ConfigOptions.key("table-name")
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription("Table name of the SqlServer database to monitor.");
+
```
Comment (ashulin) [inline flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableFactory.java:79]: Missing startup mode options and processing code.

**Review by ashulin — COMMENTED** [review]
I left some comments.

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableFactory.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+import org.apache.flink.table.utils.TableSchemaUtils;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.ververica.cdc.debezium.table.DebeziumOptions.DEBEZIUM_OPTIONS_PREFIX;
+import static com.ververica.cdc.debezium.table.DebeziumOptions.getDebeziumProperties;
+
+/** Factory for creating configured instance of {@link SqlServerTableSource}. */
+public class SqlServerTableFactory implements DynamicTableSourceFactory {
```
Comment (PatrickRen) [inline flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableFactory.java:38]: Can we have an option for configuring server time zone? I think it's necessary for configuring `RowDataDebeziumDeserializeSchema`

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableSource.java [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    // --------------------------------------------------------------------------------------------
+
+    /** Data type that describes the final output of the source. */
+    protected DataType producedDataType;
+
+    /** Metadata that is appended at the end of a physical source row. */
+    protected List<String> metadataKeys;
+
+    public SqlServerTableSource(
+            TableSchema physicalSchema,
```
Comment (PatrickRen) [inline flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableSource.java:75]: Looks like `TableSchema` has been marked as deprecated. What about using `ResolvedSchema` instead?

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/SqlServerSource.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+                    break;
+                case SCHEMA_ONLY:
+                    props.setProperty("snapshot.mode", "schema_only");
+                    break;
+                default:
+                    throw new UnsupportedOperationException();
+            }
+
+            if (dbzProperties != null) {
+                dbzProperties.forEach(props::put);
```
Comment (PatrickRen) [inline flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/SqlServerSource.java:147]: nit: `props.putAll(dbzProperties)`

**Review by PatrickRen — COMMENTED** [review]
Thanks for the contribution @GOODBOY008 ! Very excited to have this new connector in Flink CDC. I left some comments in this PR.

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableSource.java (hunk 7 above)
Comment (GOODBOY008) [reply flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableSource.java:75]: @PatrickRen Thank you for your advice! Should I create issue and submit pr for other connectors?

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableFactory.java (hunk 5 above)
Comment (GOODBOY008) [reply flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableFactory.java:78]: Hi, @ashulin i have added. please cc.

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableSource.java (hunk 7 above)
Comment (PatrickRen) [reply flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableSource.java]: Yes it's great to get rid of deprecated APIs

Comment (leonardBang) [reply flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableSource.java]: Creating a new issue and improve all connectors in another PR makes sense to me.

File: flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableFactory.java (hunk 6 above)
Comment (GOODBOY008) [reply flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableFactory.java:37]: > Can we have an option for configuring server time zone? I think it's necessary for configuring `RowDataDebeziumDeserializeSchema`
Yes,I will add server time zone option and configure `RowDataDebeziumDeserializeSchema`.

Comment (GOODBOY008) [reply flink-connector-sqlserver-cdc/src/main/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerTableFactory.java:37]: Hi,@PatrickRen.Currently,Debezium does not support Spatial column types,according to class `io.debezium.connector.sqlserver.SqlServerValueConverters`.Is there need to create class `SqlServerDeserializationConverterFactory` and configure `RowDataDebeziumDeserializeSchema`'s `userDefinedConverterFactory`?

**Review by waywtdcc — APPROVED** [review]

File: flink-connector-sqlserver-cdc/src/test/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerConnectorITCase.java [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+                                + " 'password' = '%s',"
+                                + " 'database-name' = '%s',"
+                                + " 'table-name' = '%s'"
+                                + ")",
+                        MSSQL_SERVER_CONTAINER.getHost(),
+                        MSSQL_SERVER_CONTAINER.getMappedPort(MS_SQL_SERVER_PORT),
+                        MSSQL_SERVER_CONTAINER.getUsername(),
+                        MSSQL_SERVER_CONTAINER.getPassword(),
+                        "inventory",
+                        "dbo.products");
```
Comment (leonardBang) [inline flink-connector-sqlserver-cdc/src/test/java/com/ververica/cdc/connectors/sqlserver/table/SqlServerConnectorITCase.java:89]: I think we need introduce `schemaName` option and then make the `tableName` option value to `products` for unification consideration.

**Review by leonardBang — COMMENTED** [review]
Thanks @GOODBOY008 for the contribution, looks generally good to me, I left some comments

**Review by leonardBang — APPROVED** [review]
Thanks @GOODBOY008 for the great work, LGTM.

**Pull request merged**